core/repo_status.py

```diff
@@ -0,0 +1,249 @@
+
+    def on_post_save_async(self, view):
+        global active_view
+        if active_view and active_view.buffer_id() == view.buffer_id():
+            maybe_update_status_for_view(active_view)
+        else:
+            maybe_update_status_for_view(view)
+
+
+class gs_update_status_bar(TextCommand):
+    def run(self, edit):
+        sublime.set_timeout_async(partial(maybe_update_status_for_view, self.view))
+
+
+def maybe_update_status_for_view(view):
+    """Record intent to update the status bar."""
+    if view_is_transient(view):
+        return
+
+    repo_path = repo_path_for_view(view)
+    if repo_path:
+        maybe_update_status_async(repo_path)
+
+
+def view_is_transient(view: sublime.View) -> bool:
+    """Return whether a view can be considered 'transient'.
+
+    For our purpose, transient views are 'detached' views or widgets
+    (aka quick or input panels).
+    """
+
+    # 'Detached' (already closed) views don't have a window.
+    window = view.window()
+    if not window:
+        return True
+
+    # Widgets are normal views but the typical getters don't list them.
+    group, index = window.get_view_index(view)
+    if group == -1:
+        return True
+
+    return False
+
+
+def maybe_update_status_async(repo_path):
+    # type: (RepoPath) -> None
+    with _lock:
+        current_token[repo_path] = token = uuid.uuid4().hex
+    sink = partial(update_status, repo_path)
+    pred = lambda: current_token.get(repo_path) == token
+    sublime.set_timeout_async(
+        partial(executor, pred, sink)
+    )
+
+
+def update_status(repo_path):
+    # type: (RepoPath) -> None
+    invalidate_token(repo_path)
+
+    git = make_git(repo_path)
+    status = fetch_status(git)
+    state.update_state(repo_path, status)
+
+
+def executor(pred, sink):
+    # type: (Thunk, Thunk) -> None
+    if pred():
+        sink()
+
+
+def invalidate_token(repo_path):
+    # type: (RepoPath) -> None
+    with _lock:
+        current_token.pop(repo_path, None)
+
+
+def make_git(repo_path):
+    # type: (RepoPath) -> GitCommand
+    git = GitCommand()
+    git.repo_path = repo_path
+    return git
+
+
+def fetch_status(git):
+    # type: (GitCommand) -> RepoStatus
+    first_line, *rest = git._get_status()
+    rebasing = git.in_rebase()
+    rebase_branch_name = git.rebase_branch_name() if rebasing else None
+    merging = git.in_merge()
+    merge_head = git.merge_head() if merging else None
+
+    info = {
+        'clean': len(rest) == 0,
+        'rebasing': rebasing,
+        'rebase_branch_name': rebase_branch_name,
+        'merging': merging,
+        'merge_head': merge_head
+    }  # type: RepoStatus
+
+    info.update(parse_first_line(first_line))
+    file_statuses = info['file_statuses'] = parse_file_statuses(rest)
+    info.update(group_status_entries(file_statuses))
+
+    info['short_status'] = format_short_status(info)
+
+    return info
+
+
+def format_short_status(info):
+    # type: (RepoStatus) -> str
```

Comment:
'short_status' is derived state for the `status_bar_view`, the formatting can thus be moved to `status_bar_view`